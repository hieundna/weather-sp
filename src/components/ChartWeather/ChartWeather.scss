.weather-chart {
    background-color: #fff;
    position: relative;
    max-width: 100%;
    height: 40vh;
    overflow: auto;

    .chart-type {
        position: absolute;
        top: 5px;
        left: 10px;

        span {
            padding: 3px 5px;
        }

        :nth-child(1) {
            color: lightseagreen;
        }

        :nth-child(2) {
            color: gray;
        }

        :nth-child(3) {
            color: orange;
        }
    }

    .custom-tooltip {
        color: lightseagreen;
        background-color: white;
        border: 1px solid rgba(0, 0, 0, 0.5);
        padding: 0px 15px;
    }

    .recharts-label {
        color: orange;
        font-size: 20px;
        font-weight: 600;
        bottom: 30px;
    }
}