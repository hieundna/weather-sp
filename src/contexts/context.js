import { createContext, useCallback, useContext, useReducer } from 'react';
import { rootReducer, initialState } from './reducers/rootReducer'

const Context = createContext({});

export const AppProvider = props => {
    const { children } = props;
    const [data, dispatch] = useReducer(rootReducer, initialState);
    return (
        <Context.Provider value={{ data, setData: dispatch }}>
            {children}
        </Context.Provider>
    )
}

export const useAppContext = () => {
    const { data, setData } = useContext(Context)

    const dispatch = useCallback(
        ({ type, payload }) => {
            setData({ payload, type })
        },
        [setData]
    )
    return [data, dispatch]
}

export default Context;