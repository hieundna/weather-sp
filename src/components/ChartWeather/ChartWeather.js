import '../../styles/ChartWeather.scss';
import { useAppContext } from '../../contexts/context'
import { CustomTooltip, CustomizedActiveDot } from '../../utils/utils'
import { useEffect, useState } from 'react';
import { AreaChart, ResponsiveContainer, Label, Area, ReferenceArea, CartesianGrid, XAxis, Tooltip } from 'recharts';

function ChartWeather() {
    const [{ chartData }] = useAppContext();
    const [nightTime, setNightTime] = useState([{ start: 0, end: 3 }]);
    const [windowWidth, setWindowWidth] = useState(window.innerWidth);

    useEffect(() => {
        const night = parseInt(chartData.length / 12);
        let arr = [{ start: 0, end: 3 }];
        if (night > 0) {
            for (let i = 0; i < night; i++) {
                if (i !== night - 1) {
                    arr.push({ start: arr[i].end + 6, end: arr[i].end + 12 })
                }
                else {
                    arr.push({ start: arr[i].end + 6, end: arr[i].end + 9 })
                }
            }
        }
        console.log('NightTime', arr);
        setNightTime([...arr]);
        onResizeWidth();

        window.addEventListener('resize', onResizeWidth);
        return () => {
            window.removeEventListener('resize', onResizeWidth);
        }
    }, [])

    useEffect(() => {
        if (nightTime.length > 1) {
            setTimeout(() => {
                setWindowWidth(windowWidth + 1);
                onResizeWidth();
            }, 500)
        }
    }, [nightTime])


    const onResizeWidth = e => {
        const width = window.innerWidth;
        setWindowWidth(width)
    };
    return (
        <div className='weather-chart'>
            <ResponsiveContainer key={windowWidth} width={2500} height='100%'>
                <AreaChart data={chartData}>
                    <XAxis dataKey="hour" />
                    <Tooltip content={<CustomTooltip />} />
                    <CartesianGrid stroke="#ccc" strokeDasharray="0 5" />
                    <Area type="monotone" dataKey="tide" stroke="#8884d8" fillOpacity={1} fill="#94d6f7" />
                    {nightTime ? nightTime.map((item, idx) => (
                        <ReferenceArea key={idx} x1={item.start} x2={item.end} y1={0} y2={8} strokeOpacity={1} >
                            {
                                idx !== nightTime.length - 1
                                && <Label
                                    value={`Day ${idx + 1}`}
                                    offset={50}
                                    position="right"
                                />
                            }
                        </ReferenceArea>
                    )) : null}
                    <Area
                        type="monotoneX"
                        dataKey="sun"
                        stroke="orange"
                        fillOpacity={0}
                        dot={windowWidth <= 1024 ? <CustomizedActiveDot /> : false}
                        activeDot={<CustomizedActiveDot />}
                    />
                </AreaChart>
            </ResponsiveContainer>
            <div className='chart-type'>
                <span>Tide</span>
                <span>â€¢</span>
                <span>Sunrire & Sunset</span>
            </div>
        </div>
    );
}

export default ChartWeather;
