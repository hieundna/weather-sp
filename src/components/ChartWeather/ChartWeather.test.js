import React from 'react';
import ChartWeather from './ChartWeather.js';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AppProvider } from '../../contexts/context'

const wrapper = mount(<AppProvider> <ChartWeather /> </AppProvider>);
describe('ChartWeather Component', () => {
    test("expect to render ChartWeather component", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    })
    test('renders without error', () => {
        const chartWeatherComponent = wrapper.find(".chart-type");
        expect(chartWeatherComponent.length).toBe(1);
    });

})