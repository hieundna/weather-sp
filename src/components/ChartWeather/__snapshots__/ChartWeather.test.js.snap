// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartWeather Component expect to render ChartWeather component 1`] = `
<AppProvider>
   
  <ChartWeather>
    <div
      className="weather-chart"
    >
      <ResponsiveContainer
        debounce={0}
        height="100%"
        key="1024"
        width={2500}
      >
        <ResizeDetector
          handleHeight={true}
          handleWidth={true}
          onResize={[Function]}
          targetRef={
            Object {
              "current": <div
                class="recharts-responsive-container"
                style="width: 2500px; height: 100%;"
              />,
            }
          }
        >
          <div
            className="recharts-responsive-container"
            style={
              Object {
                "height": "100%",
                "maxHeight": undefined,
                "minHeight": undefined,
                "minWidth": undefined,
                "width": 2500,
              }
            }
          >
            <AreaChart
              barCategoryGap="10%"
              barGap={4}
              data={
                Array [
                  Object {
                    "hour": "12 a.m",
                    "sun": 0,
                    "tide": 2.9,
                  },
                  Object {
                    "hour": "2 a.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "4 a.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "6 a.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 a.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "10 a.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "12 p.m",
                    "sun": 4.1,
                    "tide": 0.7,
                  },
                  Object {
                    "hour": "2 p.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "4 p.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "6 p.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 p.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "10 p.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "12 a.m",
                    "sun": 0,
                    "tide": 2.9,
                  },
                  Object {
                    "hour": "2 a.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "4 a.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "6 a.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 a.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "10 a.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "12 p.m",
                    "sun": 4.1,
                    "tide": 0.7,
                  },
                  Object {
                    "hour": "2 p.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "4 p.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "6 p.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 p.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "10 p.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "12 a.m",
                    "sun": 0,
                    "tide": 2.9,
                  },
                  Object {
                    "hour": "2 a.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "4 a.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "6 a.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 a.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "10 a.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "12 p.m",
                    "sun": 4.1,
                    "tide": 0.7,
                  },
                  Object {
                    "hour": "2 p.m",
                    "sun": 3.4,
                    "tide": 1,
                  },
                  Object {
                    "hour": "4 p.m",
                    "sun": 1.6,
                    "tide": 1.5,
                  },
                  Object {
                    "hour": "6 p.m",
                    "sun": 0.001,
                    "tide": 2,
                  },
                  Object {
                    "hour": "8 p.m",
                    "sun": 0,
                    "tide": 2.5,
                  },
                  Object {
                    "hour": "10 p.m",
                    "sun": 0,
                    "tide": 2.8,
                  },
                  Object {
                    "hour": "12 a.m",
                    "sun": 0,
                    "tide": 2.9,
                  },
                ]
              }
              height={0}
              layout="horizontal"
              margin={
                Object {
                  "bottom": 5,
                  "left": 5,
                  "right": 5,
                  "top": 5,
                }
              }
              reverseStackOrder={false}
              stackOffset="none"
              syncMethod="index"
              width={2500}
            />
          </div>
        </ResizeDetector>
      </ResponsiveContainer>
      <div
        className="chart-type"
      >
        <span>
          Tide
        </span>
        <span>
          â€¢
        </span>
        <span>
          Sunrire & Sunset
        </span>
      </div>
    </div>
  </ChartWeather>
   
</AppProvider>
`;
